### ISSUE TITLE
Doesn't pass along mouse actions

### ISSUE THREAD

**Issue by kergoth** [mobile-shell/mosh#101]
When using an xterm-compatible terminal, one can use the mouse in some applications (e.g. tmux, vim). Mosh does not pass this along, but ssh does. This is a blocking issue for me, and I won't be switching from pure ssh to mosh until/if this is resolved.

**Comment by pmccarren** [comment]
:+1: I'm blocking on this as well.

**Comment by chrisjohnson** [comment]
I've upgraded my client and server (I built it from git on OS X -- client, and I installed the ppa package on Ubuntu -- server) and I still don't see any sort of mouse support. Is there a config flag I need to enable?

Here's my server:
```
~ ❯❯❯ mosh-server

MOSH CONNECT 60003 CGI2lrlYCw9rnI0Glb/VUA

mosh-server (mosh 1.2.4a)
...
```

And my client:
```
 chrisjohnson@chris-mbp ⮀ ~ ⮀ /usr/local/bin/mosh -v
mosh 1.2.4a
...
```

And when I run /usr/local/bin/mosh my-ubuntu-server, and try mouse support, I get nothing. Vim, vim inside screen, no clicks or scrolls are picked up by any applications on the other side. When I go back to ssh and do the same thing it works great. I tried starting a new screen session and tried running vim outside of screen, nothing seems to be registered. When I scroll, it scrolls my outer iTerm scrollback, not anything on the server.

I'm using iTerm + SSH/MOSH + ccreen + vim

What am I missing?

**Comment by mamiu** [comment]
For me the same. Couldn't get it working!
Tried everything. Server and client are both running on the latest version (mosh 1.2.4a).

I'm also using iTerm2 + mosh + tmux/vim.

**Comment by andersk** [comment]
Mouse support was added in Git after the release of 1.2.4a. It is not yet part of any released version. Until there’s another release, you’ll need to compile the server and client from Git.

**Comment by chrisjohnson** [comment]
The above comment chain suggests that the ubuntu ppa should have these changes (my server), and as I mentioned I compiled from git on the client.

**Comment by niieani** [comment]
@chrisjohnson You need to test with the exact same compilation on both the client and the server, because the ppa was compiled from my branch, which now is a bit outdated, as there have been changes to the mouse support since, which might have rendered it incompatible.

**Comment by chrisjohnson** [comment]
Ok, I removed the PPA version and compiled the server from git as well. The mosh-server output is still the same (1.2.4a). I'm using the master branch in both. Same exact results, seemingly no mouse support whatsoever. Scrolling scrolls the iTerm buffer, nothing appears to be sent to the server. Clicks aren't picked up either (and I'm able to select words in vim with my mouse to be copied by iTerm -- something I can't do when mouse support is working). With and without screen.

**Comment by niieani** [comment]
@chrisjohnson What's the output of `echo $TERM` for you? Did you try a different Terminal Emulator than iTerm?

**Comment by chrisjohnson** [comment]
I have iTerm reporting xterm-256color (and $TERM mirrors that). I just tried on Terminal.app and I get the same result. Sadly I don't have a mouse to test on my local linux box, but I can see about trying PuTTY later? Though that doesn't really help since I'm obviously hoping to use my OS X terminal :P

**Comment by niieani** [comment]
@chrisjohnson @chrisjohnson 
You could also try testing with libTermKey's demo from:
http://www.leonerd.org.uk/code/libtermkey/
Compile and use: `./demo -m -p 1002`

Other than that, I really don't know what might be causing this. Are you positive that you are executing the correct compilation of mosh? Or if the server is doing so? Perhaps you have a stale package or installation hanging around that gets executed instead. That might be the most likely thing, since mouse now seemed to be working for everybody. What does `which mosh-server` do for you on the server? Does it point to the place where you `make install`ed your compiled mosh?

**Comment by mamiu** [comment]
tl;dr: The main problem was the local installation. Now there's another problem with mouse on tmux.

Ok, I could solve the problem. The reason was, that I thought mosh on the client would automatically start the mosh-server which is given by `which mosh-server` on the server. But that's not true because .bashrc, .bash_profile (and all the other shell config files) aren't executed before starting the mosh-server, so the $PATH variable isn't set to the user specific paths and therefore the old mosh-server globally installed on the server is executed.
So I have to mosh into my server with
```Shell
mosh --server="export LD_LIBRARY_PATH=/path/to/protobuf/installation/lib && /path/to/mosh/installation/mosh-server" -p 61234 user@host.com
```
after install mosh locally with this guide: https://gist.github.com/mamiu/24066ed58c158787c72f.

But now I have another problem: I can select text or panes only once, after that the mouse is disabled/locked. It behaves like the mouse outside tmux (or tmux with mouse mode disabled). This behavior doesn't occur when selecting windows at the bottom tmux bar (but after selecting text a second time I also can't switch windows with the mouse). Here's another person with the same problem: http://superuser.com/questions/882602/mouse-click-only-works-once-with-iterm2-tmux-mosh

Thanks a lot for all your help!

**Comment by chrisjohnson** [comment]
Correct @niieani, I'm positively using the right installation. I blew away old installations first, which mosh-server verifies this. There's no other mosh-server on my filesystem. I'll try to give that demo a try over the next couple days. Thanks for the help

**Comment by dickinsonmg** [comment]
I've just compiled on the client and server from the git commit c6cd99ba971123aa4072d3ab8466c3d2467db9bd version.

I'm also seeing that I'm only able to select panes once. I'm also using iTerm2 + mosh.

thanks.

**Comment by randywallace** [comment]
Actually, I just recently went thru this step (getting my client and server on the latest code) last night, via https://github.com/randywallace/mosh-portfile, and ran into this as well.  On my gateway server, I'm running on c6cd99b.   I thought it may have had something to do with the fact that I'm running nested tmux servers on several remote hosts (even tho everything works fine over plain-ole ssh), but seeing as how this is replicable by @dickinsonmg without my complicated setup, I am less inclined to think that is the issue.

I compiled and ran the demo over mosh to see what happened:

```
$ ./demo -m -p 1002
Mouse mode active
<MousePress(4)> at line=47, col=130
<MousePress(4)> at line=47, col=130
<MousePress(5)> at line=47, col=130
<MousePress(5)> at line=47, col=130
<MousePress(4)> at line=47, col=130
<MousePress(4)> at line=47, col=130
<MousePress(5)> at line=47, col=130
<MousePress(5)> at line=47, col=130
<MousePress(4)> at line=51, col=236
<C-c>
Mouse mode deactivated
```

It only works initially, or works for a little while, but it usually isn't long before I have to restart my mosh client->server connection to regain behavior. For example, if I start a new connection, open vim, set mouse=a, i can scroll all day long as long as I am in vim.  Swap between insert/visual, visual select with the mouse, yank, paste, go back to visual mouse, no problem.  If, though, I exit vim, then reopen vim, set mouse=a, and try again, it no longer works.  Or sometimes it does.

Check this out:
```
~/git/libtermkey-0.17$ ./demo -m -p 1005
Mouse mode active
<MousePress(5)> at line=32, col=204
<MousePress(5)> at line=32, col=204
<MousePress(4)> at line=32, col=204
<MousePress(4)> at line=32, col=204
<MousePress(5)> at line=32, col=204
<MousePress(5)> at line=32, col=204
<MousePress(4)> at line=32, col=204
<MousePress(5)> at line=32, col=204
<C-c>
Mouse mode deactivated
~/git/libtermkey-0.17$ C`A`A`AaAaAaA`A`A`Aa
```
Note that the terminal *after* running the demo with mode 1005 throws strange characters on the terminal when I scroll the mouse.  One time, even thought the terminal was showing that, I could still open vim, set mouse=a, and scrolling still works fine.  Tmux scrolling was not working though.

BUT, in that case, I run the demo again afterward with 1002, I still get the libtermkey expected output. When I attempt to open vim afterward, though, and scroll, it won't work.  Likewise, after running 1002, If I try 1005 again, that one won't work anymore either.

Also, things occasionally get really whacky and this happens (the character between MousePress(es) changed every time I ran the demo):

```
~/git/libtermkey-0.17$ ./demo -m -p 1005
Mouse mode active
<MousePress(5)> at line=145, col=162
P
<MousePress(4)> at line=145, col=162
P
<MousePress(5)> at line=145, col=162
P
<C-c>
Mouse mode deactivated
```

Not sure what to make of all this, but I hope this helps :)  In the meantime, I'll just keep resetting my mosh connection when I need mouse support and it isn't working (woot, tmux).

**Comment by chrisjohnson** [comment]
Just wanted to comment that miraculously, without ever touching either installation, the issue has gone away and mouse support works as of today. I just went to try libtermkey and it worked great, tested in vim, worked great, and as of yet I'm not even having any inconsistencies like @randywallace. I did nothing more than mosh my-host so perhaps a reboot on host or client cleared it up. Anyway thanks

**Comment by niieani** [comment]
@chrisjohnson thanks for the update. So it turns out it's like I said - the implementation is good, the problems were with the client/server installations... @keithw Since more people have now confirmed that everything's working re. mouse support, we should be able to close this issue now.

**Comment by danielb2** [comment]
`perl -E ' print "\e[?1005h\e[?1002h" '`  no longer works with 1.2.5

So this didn't get fixed but was made worse?

**Comment by danielb2** [comment]
I made sure the mosh-server and client are both 1.2.5 but still no luck. Any idea what could be up with that?

**Comment by chrisjohnson** [comment]
The print trick should no longer be necessary. Make sure you build from source on both using the same commit. Then try manually specifying the full path to the binaries to be sure you call the built version

**Comment by danielb2** [comment]
Well, you made me double-check the versions which was good because for some reason the server did not get upgraded and was still stuck on 1.2.4. Thanks a bundle :)

**Comment by sparr** [comment]
I'm encountering this issue with 1.2.5

On Xubuntu, my mouse works fine in xterm+ssh and intermittently fails (scrolls the window buffer) with xterm+mosh.

On OSX, my mouse behaves as arrow keys in Terminal+ssh and Terminal+mosh. my mouse works as expected in iTerm2+ssh and always fails with iTerm2+mosh

**Comment by cgull** [comment]
Reports on mosh-1.2.5's mouse functionality have generally been positive, and I use mosh, iTerm2, xterm, and weechat regularly and it all works.
Are you sure you have mosh-1.2.5 installed on both clients and server?  The mouse functionality is unlikely to work right if you have an older version of mosh installed at either end.
What application are you running on the server?

**Comment by jamorales-bsft** [comment]
on mac osx el capitan when I use the scroll wheel in a mosh terminal it goes through previous command as if I was pressing the up and down arrow keys. It does not actually scroll the terminal window at all.

**Comment by timotheecour** [comment]
having the same issue:

ssh+tmux+mouse works
mosh+tmux+mouse doesn't work

mosh version=1.2.6 on both client and server

**Comment by andersk** [comment]
@jamorales-bsft, @timotheecour: You are probably seeing #640 regarding iTerm2’s deep tmux integration (which, for reasons described there, only works over SSH), rather than this issue. You can get some level of mouse support in tmux without iTerm2’s deep integration by turning on tmux [mouse mode](http://man.openbsd.org/OpenBSD-current/man1/tmux.1#MOUSE_SUPPORT) (tmux < 2.1: `set -g mode-mouse on`, tmux ≥ 2.1: `set -g mouse on`).
